homeassistant/components/fjaraskupan/__init__.py

```diff
@@ -0,0 +1,124 @@
+"""The Fjäråskupan integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+import logging
+
+from bleak import BleakScanner
+from bleak.backends.device import BLEDevice
+from bleak.backends.scanner import AdvertisementData
+from fjaraskupan import DEVICE_NAME, Device, State
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DISPATCH_DETECTION, DOMAIN
+
+PLATFORMS = ["fan"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class EntryState:
+    """Store state of config entry."""
+
+    scanner: BleakScanner
+    device: Device
+    coordinator: DataUpdateCoordinator[State]
+    device_info: DeviceInfo
+
+
+async def async_startup_scanner(hass):
+    """Startup the background event scanner."""
+    scanner = BleakScanner()
+    await scanner.start()
+
+    async def detection_callback(
+        device: BLEDevice, advertisement_data: AdvertisementData
+    ):
+        if device.name == DEVICE_NAME:
+            _LOGGER.debug(
+                "Detection: %s %s - %s", device.name, device, advertisement_data
+            )
+            async_dispatcher_send(hass, DISPATCH_DETECTION, device, advertisement_data)
+
+    scanner.register_detection_callback(detection_callback)
+    return scanner
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Fjäråskupan from a config entry."""
+    address: str = entry.data["address"]
+    ble_device = await BleakScanner.find_device_by_address(address)
+    if ble_device is None:
+        raise ConfigEntryNotReady("Can't find device")
+
+    scanner = await async_startup_scanner(hass)
+    try:
+        device = Device(ble_device)
+
+        async def async_update_data():
```

Comment:
Anything can really throw. This is not for handling or logging the exception.  It's for making sure we clean up the scanner if anything failed, to avoid system leaks.

I would have used a context manager preferable. But can't since if all goes well i need tokeep the context.

Reply:
We don't need this large scope for guarding against leaks. If we encounter an unexpected exception we'll fix bug that at that time. For now we should just handle the known places where calls may raise.